### ISSUE TITLE
Apparent delay in component removal

### ISSUE THREAD

**Issue by luetkemj** [ddmills/geotic#55]
Having a weird experience with geotic. My FOV system seems to be demonstrating a delay in component removal. 

On each frame an FOV algorithm runs and returns a Set of grid positions within view. In the FOV system, the InFov component is removed from all entities returned from an InFovQuery. An InFov component is then added to all entities in any of the grid positions returned from the FOV algorithm.

The problem can be seen in this gif:

![fov-entity](https://user-images.githubusercontent.com/925980/153861018-29b1ba25-4640-4689-9b2d-a362ec3b2cc5.gif)

It's as if the InFov components aren't being removed immediately.

Interestingly, if I forgo the InFov component removal & addition pattern and just render entities that are in any of the grid positions from the FOV algorithm, it seems to work just fine - as can be seen in this gif:

![fov-no-entity](https://user-images.githubusercontent.com/925980/153861048-9e3403e4-b05b-4325-b0cb-65b42be544a5.gif)

I'm not 100% sure yet if this is geotic related but thought I would pass it along. This pattern is something I've used before with earlier versions of geotic and other ecs libraries. I've never see this behavior before.

Thanks for all your work on this project - I audited some other ecs libraries and they just weren't anywhere near as nice to work with.

If you're curious, links to repro cases for both code paths described above.

InFov component pattern (repros delay)
https://github.com/luetkemj/skrimshank/tree/fov-entity

Without InFov component pattern (works but not an ideal solution)
https://github.com/luetkemj/skrimshank/tree/fov-no-entity

**Comment by ddmills** [comment]
i have a hypothesis. When we remove components from an entity, the query cache is updated in-place. When removing a item from an array, we change it's length, which changes how iteration works. [This stack overthrow thread documents the problem](https://stackoverflow.com/questions/24812930/how-to-remove-element-from-array-in-foreach-loop). We can confirm this by running the query again:

```js
fovQuery.get().forEach((entity) => entity.remove(entity.inFov));

console.log(fovQuery.get().length);
```

**Comment by ddmills** [comment]
This commit changed it from a `Set` to an array.

https://github.com/ddmills/geotic/commit/671db1eb6ed7fd32783279931b1fd5cfa598720a#diff-b81e21d3e4470b402859d54cd85bc06c9805b250abbc92180ab09dc933a35899L78

Previously, `query.get()` returned a clone of the results (which made this iteration/removal possible). Now it's referencing the query's cache object directly, for performance reasons.

I think having the query return a clone is more intuitive and predictable. This should be easy to reproduce with [a unit test](https://github.com/ddmills/geotic/blob/master/tests/unit/Query.spec.js)

**Comment by luetkemj** [comment]
That's makes a lot of sense actually. Spreading the results of the query into a new array before iteration fixes the issue - not sure what sort of a perf hit I might take for that but good enough for now.

```diff
- fovQuery.get().forEach((entity) => {
+ [...fovQuery.get()].forEach((entity) => {
```

Is this something you think you'll fix in a new release?

**Comment by luetkemj** [comment]
A follow on q regarding performance - is it faster to update a query by adding/removing components from a small set of entities or iterate through a large number of entities testing against component properties (but never removing the components)?

Specifically thinking about the add/remove InFov pattern above or just adding an InFov component to everything and checking each entity for some property on the component.

Regardless, the calculation would have to occur every frame.

**Comment by ddmills** [comment]
@luetkemj yes, i definitely want to fix this in a future release. I think cloning the array from the `.get()` is worth the performance hit since it would remove confusion. Perhaps an config option to disable it?

```js
const engine = new Engine({ 
    cloneQueryResults: false
});
```

Regarding performance, i think both approaches are fine and might be similar in speed. it likely depends on specifics. it would be an interesting test! For FOV, i have [done it the same as you](https://github.com/ddmills/sleepy/blob/master/src/systems/FOVSystem.js), where you loop over a small set and operate on that. Mine updates whenever a character/player takes an action

**Comment by luetkemj** [comment]
Config option sounds good to me - if possible it might be nice to add the config option on the individual queries instead. Probably wouldn't matter as much on queries that don't update as often.

**Comment by ddmills** [comment]
@luetkemj this is fixed in `geotic@4.2.0` 😎
